src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -128,330 +147,392 @@ impl<T: 'static> EventLoop<T> {
         Ok(event_loop)
     }
 
-    pub fn run<F>(mut self, callback: F) -> !
+    pub fn run_ondemand<F>(&mut self, mut event_handler: F) -> Result<(), ExternalError>
     where
-        F: FnMut(Event<'_, T>, &RootEventLoopWindowTarget<T>, &mut ControlFlow) + 'static,
+        F: FnMut(event::Event<'_, T>, &event_loop::EventLoopWindowTarget<T>, &mut ControlFlow),
     {
-        let exit_code = self.run_return(callback);
-        process::exit(exit_code);
+        if self.loop_running {
+            return Err(ExternalError::AlreadyRunning);
+        }
+
+        loop {
+            if let PumpStatus::Exit(code) = self.pump_events_with_timeout(None, &mut event_handler)
+            {
+                if code == 0 {
+                    break Ok(());
+                } else {
+                    break Err(ExternalError::ExitFailure(code));
+                }
+            }
+        }
     }
 
-    pub fn run_return<F>(&mut self, mut callback: F) -> i32
+    pub fn pump_events<F>(&mut self, event_handler: F) -> PumpStatus
     where
-        F: FnMut(Event<'_, T>, &RootEventLoopWindowTarget<T>, &mut ControlFlow),
+        F: FnMut(event::Event<'_, T>, &event_loop::EventLoopWindowTarget<T>, &mut ControlFlow),
     {
-        let mut control_flow = ControlFlow::Poll;
+        self.pump_events_with_timeout(Some(Duration::ZERO), event_handler)
+    }
 
-        // XXX preallocate certian structures to avoid allocating on each loop iteration.
-        let mut window_ids = Vec::<WindowId>::new();
-        let mut compositor_updates = Vec::<WindowCompositorUpdate>::new();
-        let mut buffer_sink = EventSink::new();
+    fn pump_events_with_timeout<F>(
+        &mut self,
+        timeout: Option<Duration>,
+        mut callback: F,
+    ) -> PumpStatus
+    where
+        F: FnMut(event::Event<'_, T>, &event_loop::EventLoopWindowTarget<T>, &mut ControlFlow),
+    {
+        if !self.loop_running {
+            self.loop_running = true;
 
-        callback(
-            Event::NewEvents(StartCause::Init),
-            &self.window_target,
-            &mut control_flow,
-        );
+            // Reset the internal state for the loop as we start running to
+            // ensure consistent behaviour in case the loop runs and exits more
+            // than once
```

Comment:
All the comments (doc non doc) inside the Wayland backend start with capital latter and end with `.`. I think it's a generally that way in winit, though I haven't forced it there.